include/pressio/ops/eigen/ops_add_to_diagonal.hpp

```diff
@@ -49,7 +49,34 @@
 #ifndef OPS_EIGEN_OPS_ADD_TO_DIAGONAL_HPP_
 #define OPS_EIGEN_OPS_ADD_TO_DIAGONAL_HPP_
 
-namespace pressio{ namespace ops{
+namespace pressio{
+
+namespace impl{
+
+template<class T>
+bool eigen_entry_is_null(const T &M, int row, int col) {
+  // adapted from Eigen's lower_bound method (SpareCompressedBase.h line 182)
+  auto outer = M.IsRowMajor ? row : col;
+  auto inner = M.IsRowMajor ? col : row;
+  auto start = M.outerIndexPtr()[outer];
+  auto end = M.isCompressed() ? M.outerIndexPtr()[outer + 1]
+                              : M.outerIndexPtr()[outer] + M.innerNonZeroPtr()[outer];
+  auto it = std::lower_bound(M.innerIndexPtr() + start, M.innerIndexPtr() + end, inner);
+  bool found = (it != M.innerIndexPtr() + end) && (*it == inner);
+
+  return !found;
+}
```

Comment:
why is this so complicated?

Reply:
looking online, I see: 
https://eigen.tuxfamily.org/dox-devel/classEigen_1_1SparseMatrix.html#ae6f4db56c76e4d374eb8507fbef5bdb5

```
Eigen::SparseMatrix M(3,3);
auto d = M.diagonal();
for (int i=0; i<M.rows(); i++) {
  d(i) = 1;
 }
 ```
 this should work if all eleemnts exist, otherwise raise an assert

Reply:
<img width="867" alt="image" src="https://github.com/Pressio/pressio/assets/18708772/70da48fa-06e5-4cf8-a6f8-c6abbc2dd9da">

Reply:
I've updated the function to use this logic instead. The only drawback I see is that the function can't take M in by reference, since that would mess up any entries that were already set. Let me know what you think.